src/Feather.jl

```diff
@@ -46,35 +44,36 @@ juliastoragetype(meta::Metadata.TimestampMetadata, values_type) = Arrow.Timestam
 juliastoragetype(meta::Metadata.DateMetadata, values_type) = Arrow.Date
 juliastoragetype(meta::Metadata.TimeMetadata, values_type) = Arrow.Time{TimeUnit2julia[meta.unit]}
 
-juliatype{T<:Arrow.Timestamp}(::Type{T}) = DateTime
-juliatype{T<:Arrow.Date}(::Type{T}) = Date
-juliatype{T}(::Type{T}) = T
+juliatype(::Type{<:Arrow.Timestamp}) = DateTime
+juliatype(::Type{<:Arrow.Date}) = Date
+juliatype(::Type{T}) where {T} = T
 juliatype(::Type{Arrow.Bool}) = Bool
 
-addlevels!{T}(::Type{T}, levels, orders, i, meta, values_type, data, version) = return
-function addlevels!{T <: CategoricalValue}(::Type{T}, catlevels, orders, i, meta, values_type, data, version)
-    ptr = pointer(data) + meta.levels.offset
-    offsets = unsafe_wrap(Array, convert(Ptr{Int32}, ptr), meta.levels.length + 1)
+addlevels!(::Type{T}, levels, orders, i, meta, values_type, data, version) where {T} = return
+function addlevels!(::Type{T}, catlevels, orders, i, meta, values_type, data, version) where {T <: CategoricalValue}
+    ptr = convert(Ptr{Int32}, pointer(data) + meta.levels.offset)
+    offsets = [unsafe_load(ptr, i) for i = 1:meta.levels.length + 1]
     ptr += getoutputlength(version, sizeof(offsets))
-    catlevels[i] = map(x->unsafe_string(ptr + offsets[x], offsets[x+1] - offsets[x]), 1:meta.levels.length)
+    ptr2 = convert(Ptr{UInt8}, ptr)
+    catlevels[i] = map(x->unsafe_string(ptr2 + offsets[x], offsets[x+1] - offsets[x]), 1:meta.levels.length)
     orders[i] = meta.ordered
     return
 end
 
-schematype{T}(::Type{T}, nullcount, nullable, wrs) = (nullcount == 0 && !nullable) ? Vector{T} : NullableVector{T}
-schematype{T <: AbstractString}(::Type{T}, nullcount, nullable, wrs) = wrs ? NullableVector{WeakRefString{UInt8}} : NullableVector{String}
-schematype{T, R}(::Type{CategoricalValue{T, R}}, nullcount, nullable, wrs) = (nullcount == 0 && !nullable) ? CategoricalVector{T, R} : NullableCategoricalVector{T, R}
+schematype(::Type{T}, nullcount, nullable, wrs) where {T} = ifelse(nullcount == 0 && !nullable, T, Union{T, Null})
+schematype(::Type{<:AbstractString}, nullcount, nullable, wrs) = (s = ifelse(wrs, WeakRefString{UInt8}, String); return ifelse(nullcount == 0 && !nullable, s, ?s))
+schematype(::Type{CategoricalValue{T, R}}, nullcount, nullable, wrs) where {T, R} = ifelse(nullcount == 0 && !nullable, CategoricalValue{T, R}, CategoricalValue{Union{T, Null}, R})
```

Comment:
`CategoricalValue{Union{T, Null}, R}` should be `Union{CategoricalValue{T, R}, Null}`. I hope the fact that it isn't `<: CategoricalValue` isn't an issue?